### ISSUE TITLE
Cider cannot open Jar files by SSHx/Tramp

### ISSUE THREAD

**Issue by igrishaev** [clojure-emacs/cider#3409]
## Expected behavior

I have a machine that runs an nREPL server. I can reach the files in Emacs by running this:

```
(dired "/sshx:foobar@169.254.117.51:~/work/XXX")
```

I open `deps.edn` file and perform `cider-connet` to the remote port and then `cider-ns-refresh`. Navigation by the files located in `src` works great: I can jump to definitions with `M-.` with ease. But when I try to open a namespace that comes from `~/.m2`, say `clojure.core`, the following happens.

- Emacs hangs for a while transferring the file; I can see a message like:
```
Tramp: Checking ‘vc-registered’ for /sshx:user@169.254.117.51:/Users/foo/.m2/repository/org/clojure/clojure/1.10.3/clojure-1.10.3.jar...done
```
- Then it shows a buffer titled `clojure-1.10.3.jar` with the following content:
```
M  ...   Filename
- ----- ---------------
- ----- ---------------
```

- Then it reports that the search has failed:
```
cond: Search failed: "clojure/core.clj$"
```

## Actual behavior

I expected to see the actual clojure/core.clj file opened at the proper line. What's interesting here, I worked before I restarted Emacs (I've have had it running for a long time).

## Steps to reproduce the problem

The second way to reproduce this is to run `M-x cider-find-ns RET clojure.core RET` and get the same buffer and a message saying the search has failed.

### CIDER version information

```
;; CIDER 1.7.0 (Côte d'Azur), nREPL 1.0.0
;; Clojure 1.10.3, Java 20.0.1
```

### Lein / Clojure CLI version

Clojure 1.10.3
Clojure CLI version 1.11.1.1347 

### Emacs version

GNU Emacs 28.1 (build 1, aarch64-apple-darwin21.1.0, NS appkit-2113.00 Version 12.0.1 (Build 21A559)) of 2022-05-11

### Operating system

MacOS 12.3.1

### JDK distribution

OpenJDK 20.0.1, no docker

**Comment by vemv** [comment]
Thanks!

cider (master, must be greater than 1.7.0) and cider-nrepl 0.35.0 improve a number of things in the xref front.

Although I don't use tramp, I wouldn't be surprised if this started working, please give it a try.

**Comment by igrishaev** [comment]
Sure, let me try...

**Comment by igrishaev** [comment]
@vemv no it didn't help: I've installed the latest master (Cider 1.8.0-snapshot) and cider-nrepl 0.35.0 but the jar behaviour is still the same.

**Comment by igrishaev** [comment]
One more interesting thing: if I step into the buffer and press `g`, for a moment it's get filled with machinery symbolds, and then it becomes this:
```
M    Mode     Size        Date&time         Filename
- ----------  -----  --------------------  ----------------
  -rw-rw-rw-    244  18-Jul-2023 12:23:36  META-INF/MANIFEST.MF
  drwxrwxrwx      0  18-Jul-2023 12:20:26  lambda/
  -rw-rw-rw-   1306  18-Jul-2023 12:19:06  lambda/aes.clj
  -rw-rw-rw-    901  18-Jul-2023 12:19:06  lambda/core.clj
  drwxrwxrwx      0  18-Jul-2023 12:20:26  lambda/test/
  drwxrwxrwx      0  18-Jul-2023 12:20:26  lambda/test/fixture/
  -rw-rw-rw-   1542  18-Jul-2023 12:19:06  lambda/test/fixture/core.clj
  -rw-rw-rw-    913  18-Jul-2023 12:19:06  
...
```

**Comment by igrishaev** [comment]
If I open the `lambda/core.clj` entry and try to `M-x` to `lambda.core`, it works great: Emacs navigates to the proper file.

**Comment by igrishaev** [comment]
Another update:

- when I `M-.` to a jar namespace the first time, it doesn't work: it shows the following buffer:
```
M  ...   Filename
- ----- ---------------
- ----- ---------------
```

- then I press `g` and it becomes a table of Jar entities
- if I switch back to code and navigate to any namespace from this Jar, it works but...
- ...if I try to navigate to some other namespace from this namespace (e.g. jdbc.next or somethis), it doesn't work: it asks for a path to Jar file.
-

**Comment by vemv** [comment]
> I wonder if that $ is a badly formed regex?

...probably not, it appears to work fine for a simple experiment.

@igrishaev, the main logic resides in `cider-find-file`, and specifically:

https://github.com/clojure-emacs/cider/blob/4a35e243c2e6830af23d05dcb648c71816e3da97/cider-common.el#L372-L380

> One more interesting thing: if I step into the buffer and press g, for a moment it's get filled with machinery symbolds, and then it becomes this ...

Surely the `re-search-forward` fails because nothing is displayed until the effect of pressing `g` is applied?

IDK how to automate that (IDK much about tramp in general). It would be tremendous help if you could try improving the mentioned lines (L372-L380)

**Comment by vemv** [comment]
> ..if I try to navigate to some other namespace from this namespace (e.g. jdbc.next or somethis), it doesn't work: it asks for a path to Jar file.

Could you be more specific?

**Comment by igrishaev** [comment]
@vemv sure, sorry for confusion. I mean, I can navigate to the `lambda.core` namespace (by pressing `g` in a broken buffer). I can see the ns header:
```
(ns lambda.core
  (:gen-class)
  (:require [lambda.util :as util]
            [clojure.tools.logging :as log]
            [aws.lambda :as lambda]
            [clojure.string :as str]))

```
Now I want to proceed to the `aws.lambda` namespace. I move the point to the `aws.lambda` word, press `M-.` but it either says "No linked Cider sessions" or opens a minubufer with jar files.

**Comment by vemv** [comment]
Thanks!

Could you paste/screenshot that minibuffer?

**Comment by igrishaev** [comment]
@vemv sure I've sent it to you in Slack (this is a private project).

**Comment by igrishaev** [comment]
**UPD:** the problem with Jar buffers has nothing in common with Cider. Turned out, I had at all my computers. The culprit was the following line:
```
(setq-default coding-system-for-read 'utf-8)
```
Once I removed it, jumping into the Jar files (even via Tramp) works fine.

**Comment by igrishaev** [comment]
The only problem I have now is, sometimes, stepping into a Jar files leads to loosing the current Cider session. But it looks like there is already an existing topic for that. We can close this one, I believe, if nobody minds.

**Comment by vemv** [comment]
I'll try to ensure `coding-system-for-read` is tempoarily bound for us

**Comment by vemv** [comment]
I tried to reproduce the `coding-system-for-read` issue, but couldn't.

e.g. I set it to `'binary` or `'iso-safe` to see if anything would break, but it wasn't the case.

Probably this was enough of an edge case (as most users wouldn't set this variable) to defer any fix for the time being.

btw, would love a confirmation that a workaround for session linking is [no longer needed](https://github.com/clojure-emacs/cider/pull/3412#issuecomment-1686326704).

Cheers - V
